pkg/apis/autoscaling/v1alpha1/metricsproducer_validation.go

```diff
@@ -12,9 +12,113 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// +kubebuilder:webhook:verbs=create;update,path=/validate-autoscaling-karpenter-sh-v1alpha1-metricsproducer,mutating=false,sideEffects=None,failurePolicy=fail,groups=autoscaling.karpenter.sh,resources=metricsproducers,versions=v1alpha1,name=vmetricsproducer.kb.io
 package v1alpha1
 
-import "fmt"
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/runtime"
+	"regexp"
+	"strings"
+	"time"
+)
+
+// Only Validates ScheduledCapacity MetricsProducer right now
+func (m *MetricsProducer) ValidateCreate() error {
+	if m.Spec.Schedule != nil {
+		if err := m.ValidateSchedule(); err != nil {
+			return err
+		}
+	}
+	if m.Spec.ReservedCapacity != nil {
+		return m.Spec.ReservedCapacity.validate()
+	}
+	if m.Spec.PendingCapacity != nil {
+		return m.Spec.PendingCapacity.validate()
+	}
+	if m.Spec.Queue != nil {
+		m.Spec.ValidateQueue()
+	}
+
+	return nil
+}
+
+func (m *MetricsProducer) ValidateUpdate(old runtime.Object) error {
+	return m.ValidateCreate()
+}
+
+func (m *MetricsProducer) ValidateDelete() error {
+	return nil
+}
+
+func (m *MetricsProducer) ValidateSchedule() error {
+	if m.Spec.Schedule != nil {
+		for _, b := range m.Spec.Schedule.Behaviors {
+			if err := b.Start.IsValidPattern(); err != nil {
+				return fmt.Errorf("start pattern could not be parsed, %w", err)
+			}
+			if err := b.End.IsValidPattern(); err != nil {
+				return fmt.Errorf("end pattern could not be parsed, %w", err)
+			}
+			if b.Replicas < 0 {
+				return fmt.Errorf("behavior.replicas cannot be negative")
+			}
+		}
+	}
+	if m.Spec.Schedule.DefaultReplicas < 0 {
+		return fmt.Errorf("defaultReplicas cannot be negative")
+	}
+	if m.Spec.Schedule.Timezone != nil {
+		if _, err := time.LoadLocation(*m.Spec.Schedule.Timezone); err != nil {
+			return fmt.Errorf("timezone region could not be parsed")
+		}
+	}
+	return nil
+}
+
+// These regex patterns are meant to match to one element at a time
+const (
+	weekdayRegexPattern = "^((sun(day)?|0|7)|(mon(day)?|1)|(tue(sday)?|2)|(wed(nesday)?|3)|(thu(rsday)?|4)|(fri(day)?|5)|(sat(urday)?|6))$"
+	monthRegexPattern   = "^((jan(uary)?|1)|(feb(ruary)?|2)|(mar(ch)?|3)|(apr(il)?|4)|(may|5)|(june?|6)|(july?|7)|(aug(ust)?|8)|(sep(tember)?|9)|((oct(ober)?)|(10))|(nov(ember)?|(11))|(dec(ember)?|(12)))$"
+	onlyNumbersPattern  = `^\d+$`
+)
+
+func (p *Pattern) IsValidPattern() error {
```

Comment:
this logic is repetitive, you should be able to abstract into a loop

Reply:
Thanks, I'll do that.